src/vs/workbench/parts/preferences/common/keybindingsEditorModel.ts

```diff
@@ -90,19 +90,72 @@ export class KeybindingsEditorModel extends EditorModel {
 
 	public fetch(searchValue: string, sortByPrecedence: boolean = false): IKeybindingItemEntry[] {
 		searchValue = searchValue.trim();
-		const quoteAtFirstChar = searchValue.charAt(0) === '"';
-		const quoteAtLastChar = searchValue.charAt(searchValue.length - 1) === '"';
+
+		if (this.isSourceFilterApplied(searchValue)) {
+			searchValue = this.getSourceFilterValue(searchValue);
+			searchValue = this.parseSearchValue(searchValue);
+			return this.fetchKeybindingItemsBySource(sortByPrecedence ? this._keybindingItemsSortedByPrecedence : this._keybindingItems, searchValue, this.quoteAtFirst(searchValue) && this.quoteAtLast(searchValue));
+		}
+		searchValue = this.parseSearchValue(searchValue);
+		return this.fetchKeybindingItemsByText(sortByPrecedence ? this._keybindingItemsSortedByPrecedence : this._keybindingItems, searchValue, this.quoteAtFirst(searchValue) && this.quoteAtLast(searchValue));
+	}
+
+	private isSourceFilterApplied(searchValue: string): boolean {
+		return /^@source:/i.test(searchValue);
+	}
+
+	private getSourceFilterValue(searchValue: string): string {
+		return searchValue.split('@source:')[1].trim() || '';
+	}
+
+	private quoteAtFirst(searchValue: string) {
+		return searchValue.charAt(0) === '"';
+	}
+
+	private quoteAtLast(searchValue: string) {
+		return searchValue.charAt(searchValue.length - 1) === '"';
+	}
+
+	private parseSearchValue(searchValue: string) {
+		const quoteAtFirstChar = this.quoteAtFirst(searchValue);
+		const quoteAtLastChar = this.quoteAtLast(searchValue);
 		if (quoteAtFirstChar) {
 			searchValue = searchValue.substring(1);
 		}
 		if (quoteAtLastChar) {
 			searchValue = searchValue.substring(0, searchValue.length - 1);
 		}
-		searchValue = searchValue.trim();
-		return this.fetchKeybindingItems(sortByPrecedence ? this._keybindingItemsSortedByPrecedence : this._keybindingItems, searchValue, quoteAtFirstChar && quoteAtLastChar);
+		return searchValue.trim();
+	}
+
+	private fetchKeybindingItemsBySource(keybindingItems: IKeybindingItem[], searchValue: string, completeMatch: boolean): IKeybindingItemEntry[] {
+		if (!searchValue) {
+			return keybindingItems.map(keybindingItem => ({ id: KeybindingsEditorModel.getId(keybindingItem), keybindingItem, templateId: KEYBINDING_ENTRY_TEMPLATE_ID }));
+		}
+
+		const result: IKeybindingItemEntry[] = [];
```

Comment:
Not sure if you need this code any more because this is handled by `fetchKeybindingItemsByText`

Reply:
`fetchKeybindingItemsByText` is called conditionally when no `@source:` filter is applied.
Otherwise `fetchKeybindingItemsBySource` will be called.